sinks/cortex/cortex.go

```diff
@@ -0,0 +1,226 @@
+package cortex
+
+import (
+	"bytes"
+	"context"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/sirupsen/logrus"
+	"github.com/stripe/veneur/v14"
+	"github.com/stripe/veneur/v14/samplers"
+	"github.com/stripe/veneur/v14/sinks"
+	"github.com/stripe/veneur/v14/ssf"
+	"github.com/stripe/veneur/v14/trace"
+	"github.com/stripe/veneur/v14/util"
+)
+
+const (
+	// MaxConns is the number of simultaneous connections we allow to one host
+	MaxConns = 100
+	// DefaultRemoteTimeout after which a write request will time out
+	DefaultRemoteTimeout = time.Duration(30 * time.Second)
+)
+
+// CortexMetricSink writes metrics to one or more configured Cortex instances
+// using the prometheus remote-write API. For specifications, see
+// https://github.com/prometheus/compliance/tree/main/remote_write
+type CortexMetricSink struct {
+	URL           string
+	RemoteTimeout time.Duration
+	ProxyURL      string
+	client        *http.Client
+}
+
+type CortexMetricSinkConfig struct {
+	URL           string        `yaml:"url"`
+	RemoteTimeout time.Duration `yaml:"remote_timeout"`
+	ProxyURL      string        `yaml:"proxy_url"`
+}
+
+// Create creates a new Cortex sink.
+// @see veneur.MetricSinkTypes
+func Create(
+	server *veneur.Server, name string, logger *logrus.Entry,
+	config veneur.Config, sinkConfig veneur.MetricSinkConfig,
+) (sinks.MetricSink, error) {
+	conf, ok := sinkConfig.(CortexMetricSinkConfig)
+	if !ok {
+		return nil, errors.New("invalid sink config type")
+	}
+
+	return NewCortexMetricSink(conf.URL, conf.RemoteTimeout, conf.ProxyURL)
+}
+
+// ParseConfig extracts Cortex specific fields from the global veneur config
+func ParseConfig(config interface{}) (veneur.MetricSinkConfig, error) {
+	cortexConfig := CortexMetricSinkConfig{}
+	err := util.DecodeConfig(config, &cortexConfig)
+	if err != nil {
+		return nil, err
+	}
+	if cortexConfig.RemoteTimeout == 0 {
+		cortexConfig.RemoteTimeout = DefaultRemoteTimeout
+	}
+	return cortexConfig, nil
+}
+
+// NewCortexMetricSink creates and returns a new instance of the sink
+func NewCortexMetricSink(URL string, timeout time.Duration, proxyURL string) (*CortexMetricSink, error) {
+	return &CortexMetricSink{
+		URL:           URL,
+		RemoteTimeout: timeout,
+		ProxyURL:      proxyURL,
+	}, nil
+}
+
+// Name returns the string cortex
+func (s *CortexMetricSink) Name() string {
+	return "cortex"
+}
+
+// Start sets up the HTTP client for writing to Cortex
+func (s *CortexMetricSink) Start(*trace.Client) error {
+	// Default concurrent connections is 2
+	t := http.DefaultTransport.(*http.Transport).Clone()
```

Comment:
Do we want to reuse the traced HTTP client `veneur.server.HTTPClient`? (If so, you can access this in `Create` above.)

Reply:
nice, thanks.

Reply:
I actually reverted this fix because I realised we do need to keep control of how the http client proxies for some use cases of this sink (specifically amazon managed prometheus, which will likely make use of the sigv4 proxy since we don't implement sigv4 directly here). I don't want to attach a proxy configuration to Veneur's global http client and force all sinks through the same proxy. And while we're building our own transport, I think it makes sense to keep control of the remote timeout as well, since we may want a different timeout from the global idle configuration.

Reply:
Makes sense. Thanks!